mne_bids/mne_bids.py

```diff
@@ -325,10 +380,19 @@ def _coordsystem_json(raw, unit, orient, manufacturer, fname, verbose):
         Used to define the coordinate system for the MEG sensors.
     fname : str
         Filename to save the coordsystem.json to.
+    write_mode : str, one of ('overwrite', 'error')
+        How the file should be handled if is exists already.
+        Defaults to 'overwrite'.
+        If write_mode == 'overwrite' the old file will be removed and replaced
+        with a new one.
+        If write_mode == 'error' an `OSError` will be raised.
     verbose : bool
         Set verbose output to true or false.
 
     """
+    if write_mode == 'error':
+        _check_file_exists(fname)
```

Comment:
why not actually do this in `_write_json` so that all future sidecar writers will benefit from it?

Reply:
you asked this yesterday...
https://github.com/mne-tools/mne-bids/pull/99#discussion_r224306381

Reply:
oooh, I see what you mean. I thought about that but then it would do all the processing before realising it didn't need to do any of it. I thought it better to do the check before it tries to process the data in the `Raw`.
It does seem like bundling it into `_write_json` would make it a lot cleaner though

Reply:
yeah I had the same thought as you. I think it's okay to fail a little later in that case.

what would be really neat is to have a similar function for `tsv` called `_write_tsv` ...

Reply:
I guess that would just be
```python
def _write_tsv(fname, df):
    df.to_csv(fname, sep='\t', index=False, na_rep='n/a')
```
:P
I can add that to this PR if you would like, that way we can also give it the `write_mode` as well, add `write_mode` as an argument for `_write_json`, and get rid of the `_check_file_exists` function as I don't like it anyway because of the name <-> functionality mismatch

Reply:
I was wondering if the logic of `append` can be baked into `_write_tsv` since it's duplicated right now. It would make the code a little more DRY.

yes okay, maybe we don't need `_check_file_exists` then ...

Reply:
I was thinking the arguments would look something like:
```python
def _write_tsv(fname, df, error_on_exist=False, verbose=False):
```
and it can be called with:
```python
_write_tsv(fname, df, (write_mode == 'error'), verbose)
```

The functionality from `_check_file_exists` then gets moved into this function (and the `_write_json` function)

Reply:
I think you should strive to refactor [these lines](https://github.com/mne-tools/mne-bids/blob/11465ad16a234f528cc9137e6373824a5008bcb2/mne_bids/mne_bids.py#L287-L304) and [these](https://github.com/mne-tools/mne-bids/blob/11465ad16a234f528cc9137e6373824a5008bcb2/mne_bids/mne_bids.py#L346-L362). It seems like quite a bit of duplicated code.

Reply:
Would that not be part of what would be covered by a fix for #110 ?
There is some reasonably specific info in those lines (eg. column names, which column to sort or drop duplicates of).
I feel this would be better suited as part of @sappelhoff 's idea to remove the requirement for Pandas as this merging function could be it's own separate function.

Reply:
It would be `_write_tsv(data, columns, subset, sorty_by, write_mode, verbose)`. I think it's still not so bad when considering the number of duplicated lines across 4 private functions. Also it would help in the transition to `pandas` as @sappelhoff would then have to just modify this private function.

Reply:
perhaps you could make a separate PR for this and rebase this PR against master once that is merged? That could give you a feeling of progress rather than getting stuck on my nitpicks on this PR ...

Reply:
you are also missing a `fname` argument
Also, I think this goes against the whole "each function should have one purpose", as the purpose of the `_write_tsv` function will now be "do all the constructing of the data, and merging if required", plus "write the data to a tsv file".
I don't disagree that those two chunks of code could be condensed down in some way or handled by another private function, but I think `_write_tsv` is not the function to move that processing to.

Reply:
okay fair enough. Let's start small and go with `_write_tsv` which accepts `df` as you suggest. We can add another private function later if required.